README.md
# spitblit

This is intended to be a collection of pure-Lua low-level drivers for various TFT displays over SPI for NodeMCU running on the ESP8266 (and/or possibly ESP32).

Right now, it only has support for the ILI9225, because that's all I have, but I'm trying to factor the interface to be portable to other modules.

## Version 0.1.0

This project follows Semantic Versioning.

See CHANGELOG.md for version history.

## API

Each module in the `drivers` directory implements a constructor for a screen of that kind, which will return a table exposing the `screen` interface defined below.

Drivers may provide unit-specific extensions to this interface: see the corresponding documentation in `docs`.

### Constructor

The constructor for a screen takes parameters defining the hardware interface (such as pin addresses) for the screen it will drive.

See the documentation for the specific screen for all the parameters that are required and/or optional for that screen's driver constructor.

### `screen:init()`

This method runs all initialization procedures required to set up the screen, including setting a window that encompasses the entire screen and turning on the backlight.

If you won't need to re-initialize the screen at any point, and you don't have any dangling references to the driver's constructor, you may free up the memory taken up by these initialization procedures by setting `screen.init = null` after calling it.

### `screen.width` and `screen.height`

The dimensions of the screen implemented by this driver, in pixels.

Note that this is based on the *native orientation* of the screen, as defined in the datasheet: unlike most consumer displays, your screen may be specified in portrait orientation, with a `width` that is less than its `height`. spitblit does not perform any sort of coordinate transformation (such details are left to higher-level wrappers around the spitblit interface): if you want to send rotated images to the screen, you'll want to set a window with addresses that decrease in either the `x` or `y` axis (depending on which way you rotate the screen). See the definition of `screen:window` described below (noting that this kind of rotation is what the default value of `vertical` is based around).

### `screen:window(x0, y0, x1, y1, vertical)`

This function sets the region of the screen that pixels will fill, going from `x0` to `x1` and from `y0` to `y1`. (If `x0`/`y0` is greater than `x1`/`y1`, the X/Y values will decrement; otherwise, they will increment.)

The `vertical` parameter defines whether pixels should fill up/down before they fill left/right (in other words, if the image should be rotated). by default, this is true if one axis is in descending order, and the other axis is in ascending order (if `x0` is greater than `x1` and `y0` is less than `y1`, or vice versa), and false otherwise.

Under the default behavior for `vertical`, changing the order of the `x0`/`y0` and `x1`/`y1` parameters will rotate the image: if you want to *flip* the image (eg. `x1` is less than `x0` and `y0` is less than `y1` and you want to draw a series of rows from right to left), you should specify this explicitly (in the previous example, that would be an explicit fifth parameter of `false`).

Note that setting the window does *not* set the address: if you want to start filling from the first position and your drawing code hasn't already moved you there, you will need to explicitly call `screen:jump(x0, y0)` immediately after calling this method.

### `screen:jump(x, y)`

Moves the current fill address to the given position, so that subsequent calls to `screen:fill(bytes)` will proceed starting from that position.

For example, if you wanted to implement a `setPixel` method directly on top of `spitblit` (which you cold do, though it wouldn't be as efficient as keeping an internal buffer and just redrawing the modified region of the screen), you could do it with a call to `jump` moving to that coordinate (assuming that the pixel is within the defined window), followed by a call to `fill` with the bytes of that pixel's color.

### `screen.bpc`

The color format of the screen, in terms of bits-per-channel.

Right now, the only format specified is `'r5g6b5'`, for the 16-bit RGB color used by the ILI9225. However, some other color formats that future screens are potentially expected to use are:

- `'r8g8b8'`, for 24-bit color
- `'r2g3b2'`, for 8-bit color
- `'w8'`, for 8-bit grayscale
- `'w1'` or `'k1'`, for 1-bit black-and-white (white or black high, respectively)

Screens defining color in a different format than bits-per-channel color (ie. an indexed palette) will probably have their color format defined by a different parameter: such screens are not yet supported, so such parameters are not yet specified.

### `screen:fill(bytes)`

Sends pixel data from `bytes` to the screen. This function passes its arguments to `spi.send` under the hood, so it supports all the same input formats as `spi.send` (lists and/or tables of numbers and/or strings).

### `screen:light()`

Sets the brightness of the screen backlight. Currently only supports turning the backlight on or off (by passing `true` / `1` or `false` / `0`, respectively).


CHANGELOG.md
# spitblit changelog

## v0.1.0

Initial release, with one driver (ili9225.lua) implementing five core methods and three constant properties:

- `screen:init(opts)`
- `screen:window(x0, x1, y0, y1, vertical)`
- `screen:jump(x, y)`
- `screen:fill()`
- `screen:light()`
- `screen.width`
- `screen.height`
- `screen.bpc`

### ili9225

The driver also exposes three non-interface methods for low-level use:

- `screen:initspi()`
- `screen:initgpio()`
- `screen:setreg()`

## v0.0.0

Effectively unversioned. Any file carrying a "v0.0.0" designation is from early enough in development that I hadn't settled on an interface to document.


spitblit.lua
return {
  version = '0.0.0',
  create = function(model, pins)
    local cr = model.reg.ctrl

    local rs = pins.rs
    local cs = pins.cs

    local writepin = gpio.write
    local LOW = gpio.LOW
    local HIGH = gpio.HIGH

    local send = spi.send

    local cmd = LOW
    local dat = HIGH

    local function push(cord, word)
      writepin(rs, cord)
      writepin(cs, LOW)
      send(1, word / 256, word % 256)
      writepin(cs, HIGH)
    end
    local function writereg(reg, data)
      push(cmd, reg)
      push(dat, data)
    end
    local function writeregs(regs, data)
      for i = 1, #regs do
        writereg(regs[i], data[i])
      end
    end

    local coordOrientations = {

    }

    local object = {}
    function object:init()
      node.setcpufreq(node.CPU160MHZ)

      spi.setup(1, spi.MASTER, spi.CPOL_LOW, spi.CPHA_LOW, spi.DATABITS_8, 2)

      gpio.mode(rs, gpio.OUTPUT)
      gpio.mode(pins.rst, gpio.OUTPUT)
      gpio.mode(cs, gpio.OUTPUT)
      gpio.mode(pins.led, gpio.OUTPUT)

      writepin(pins.led, HIGH)

      -- start initial sequence
      writeregs(cr.power, {0, 0, 0, 0, 0})

      -- power on
      writereg(cr.power[2], 0x0018)
      writereg(cr.power[3], 0x6121)
      writereg(cr.power[4], 0x006f)
      writereg(cr.power[5], 0x495f)
      writereg(cr.power[1], 0x0800)
      tmr.delay(10)
      writereg(cr.power[2], 0x103b)
      tmr.delay(50)

      writereg(cr.driver_output, 0x011c)
      writereg(cr.lcd_ac_driving, 0x0100)
      writereg(model.reg.entry_mode, 0x1030)
      writereg(cr.disp[1], 0x0000)
      writereg(cr.blank_period[1], 0x0808)
      writereg(cr.frame_cycle, 0x1100)
      writereg(cr.interface, 0x0000)
      writereg(cr.osc, 0x0d01)
      writereg(model.reg.vci_recycling, 0x0029)
      writereg(model.reg.set.gram_haddr, 0x0000)
      writereg(model.reg.set.gram_vaddr, 0x0000)

      writereg(cr.gate_scan, 0x0000)
      writeregs(cr.vertical_scroll, {0xdb, 0, 0})
      writeregs(model.reg.pos.partial_driving, {0xdb, 0})
      writeregs(model.reg.addr.hwin, {0xaf, 0})
      writeregs(model.reg.addr.vwin, {0xdb, 0})

      writeregs(cr.gamma,
        {0, 0x808, 0x80a, 0xa, 0xa08, 0x808, 0, 0xa00, 0x710, 0x710})

      writereg(cr.disp[1], 0x0012)
      tmr.delay(50)
      writereg(cr.disp[1], 0x1017)
    end
    return object
  end
}


drivers/ili9225.lua
--[[---
File:    drivers/ili9225.lua
From:    https://github.com/stuartpb/spitblit
Version: 0.1.0
Author:  Stuart P. Bentley <s@stuartpb.com>
License: MIT
---]]--

return function(screen)
  screen = screen or {}
  screen.pins = screen.pins or {}
  screen.pins.rs = screen.pins.rs or 2 -- GPIO4
  screen.pins.cs = screen.pins.cs or 8 -- GPIO15
  screen.pins.rst = screen.pins.rst or 4 -- GPIO2
  screen.pins.led = screen.pins.led or 1 -- GPIO5

  if screen.cpufreq == nil then
    screen.cpufreq = node.CPU160MHZ
  end

  -- fixed specifications
  screen.width = 176
  screen.height = 220

  screen.bpc = 'r5g6b5'

  -- methods
  function screen:initspi(opts)

    -- Use full frequency
    if self.cpufreq then node.setcpufreq(self.cpufreq) end

    -- Set up SPI:
    spi.setup(
      -- HSPI in Master mode
      1, spi.MASTER,
      -- Clock polarity and phase low
      spi.CPOL_LOW, spi.CPHA_LOW,
      -- 8 bits per item, so we can transmit image data from strings
      -- (spi.send doesn't know how to send multiple characters at once)
      spi.DATABITS_8,
      -- Use a clock divider of 2
      -- (40MHz, only ~100 times faster than the maximum on the datasheet :P)
      self.clock_div or 2)
  end

  function screen:initgpio()
    -- Set our pins to output

    -- Register Select pin (aka Command/Data)
    gpio.mode(self.pins.rs, gpio.OUTPUT)
    -- Chip Select pin
    gpio.mode(self.pins.cs, gpio.OUTPUT)
    -- Reset pin
    gpio.mode(self.pins.rst, gpio.OUTPUT)
    -- LED backlight pin
    gpio.mode(self.pins.led, gpio.OUTPUT)
  end

  function screen:setreg(bytes)
    local spisend = spi.send
    local gpiowrite = gpio.write
    local LOW = gpio.LOW
    local HIGH = gpio.HIGH

    local rs = self.pins.rs

    -- Select chip
    gpiowrite(self.pins.cs, LOW)

    for i = 1, #bytes, 3 do
      -- Select index register
      gpiowrite(rs, LOW)
      -- Send register index selection
      spisend(1, 0, bytes[i])

      -- Select data register
      gpiowrite(rs, HIGH)
      -- Send data bytes
      spisend(1, bytes[i + 1], bytes[i + 2])
    end

    -- Deselect chip
    gpiowrite(self.pins.cs, HIGH)
  end

  function screen:light(setting)
    if setting == 0 or setting == false or setting == gpio.LOW then
      setting = gpio.LOW
    else
      setting = gpio.HIGH
    end

    gpio.write(self.pins.led, setting)
  end

  function screen:init()
    self:initspi()
    self:initgpio()
    self:light(1)

    -- Cycle the reset pin
    gpio.write(self.pins.rst, gpio.HIGH)
    tmr.delay(1)
    gpio.write(self.pins.rst, gpio.LOW)
    tmr.delay(10)
    gpio.write(self.pins.rst, gpio.HIGH)
    tmr.delay(50)

    -- Clear the power control registers (prep for power-on sequence)
    self:setreg{
      0x10, 0, 0,
      0x11, 0, 0,
      0x12, 0, 0,
      0x13, 0, 0,
      0x14, 0, 0
    }

    -- Give the power control registers a little time to set
    tmr.delay(40)

    -- Power-on sequence (see datasheet page 104, 13.4, Figure 42)

    self:setreg{
    -- Power Control 2 (see datasheet page 64, 8.2.13)
    -- 00: Leave boost circuits off
    -- 1: Generate unamplified voltage
    -- 8: Set boost converter voltage (VCI1) to 2.58 volts
      0x11, 0x00, 0x18,

    -- Power Control 3 (see datasheet page 65, 8.2.14)
    -- 6: Set the gate voltage multipliers to 6x (high) and -4x (low)
    -- 121: Operating frequencies for the driving circuits
      0x12, 0x61, 0x21,

    -- Power Control 4 (see datasheet page 67, 8.2.15)
    -- 6f (1101111): Set gamma voltage to 4.68 volts
      0x13, 0x00, 0x6f,

    -- Power Control 5 (see datasheet page 67, 8.2.16)
    -- 49: Set panel supply high voltage to 3.76V (4.68*(.4015+.0055*0x49))
    -- 5f: Set alternating amplitude to 4.72V (4.68*(.534+.006*(0x5f-16)))
      0x14, 0x49, 0x5f,

    -- Power Control 1 (see datasheet page 63, 8.2.12)
    -- 8: Set driving capability to Medium Fast 1
      0x10, 0x08, 0x00
    }
    -- Give these settings time to propagate
    tmr.delay(10)

    -- Power Control 2 again:
    -- 10 (APON): Automatically start the boost circuits
    -- 3: Generate amplified voltage
    -- b: Set boost converter voltage (VCI1) to 2.76 volts
    self:setreg{0x11, 0x10, 0x3b}

    -- Wait for boost circuits to do their thing
    tmr.delay(50)

    -- Rest of registers

    self:setreg{
    -- Driver Output Control (see datasheet page 51, 8.2.4)
    -- 0: Normal polarity
    -- 1 (SS): Count X coordinates left-to-right
    -- 1c: Drive all 220 lines of the screen
      0x01, 0x01, 0x1c,

    -- LCD Driving Waveform Control (see datasheet page 54, 8.2.5)
    -- 0100: Single-line inversion (see http://www.techmind.org/lcd/)
      0x02, 0x01, 0x00,

    -- Entry Mode (see datasheet page 56, 8.2.6)
    -- 10 (BGR): Read in RGB order
    -- (in testing, 0 caused colors to read in BGR order, regardless of SS)
    -- 3: Increase addresses both horizontally and vertically
    -- 0: Portrait traversal (increment X addr to end before doing Y)
      0x03, 0x10, 0x30,

    -- Display Control 1 (see datasheet page 58, 8.2.7)
    -- 0000: Display off (we'll turn it on later)
      0x07, 0x00, 0x00,

    -- Display Control 2 (see datasheet page 59, 8.2.8)
    -- (The instruction table refers to this as "Blank Period Control 1")
    -- 08: 8 lines of front porch
    -- 08: 8 lines of back porch
      0x08, 0x08, 0x08,

    -- Frame Cycle Control (see datasheet page 60, 8.2.9)
    -- 1: 8 clock cycle gate delay
    -- 1: 8 clock cycle source delay
    -- 00: 16 clock cycles per line
      0x0b, 0x11, 0x00,

    -- RGB Input Interface Control 1 (see datasheet page 61, 8.2.10)
    -- 0000: We don't use the RGB Input Interface so this doesn't matter
      0x0c, 0x00, 0x00,

    -- Oscillator Control (see datasheet page 62, 8.2.11)
    -- 0d (1101): Oscillate at 444.4KHz
    -- 01: Enable oscillator
      0x0f, 0x0d, 0x01,

    -- VCI Recycling (see datasheet page 68, 8.2.17)
    -- 0020: 2 clock cycles (for what? I can't even tell)
      0x15, 0x00, 0x20,

    -- RAM Address Set (see datasheet page 69, 8.2.18)
    -- Start at 0x0000 by setting the high and low addresses to 0
      0x20, 0x00, 0x00,
      0x21, 0x00, 0x00,

    -- Gate Scan Control (see datasheet page 70, 8.2.22)
    -- 0000: start scanning from gate 1
      0x30, 0x00, 0x00,

    -- Vertical Scroll Control (see datasheet pages 71-72, 8.2.23-8.2.24)
    -- (The numbering for these in section headings is messed up)
    -- db: If we vertically scroll, do the whole screen
      0x31, 0x00, 0xdb,
      0x32, 0x00, 0x00,
      0x33, 0x00, 0x00,

    -- Partial Screen Driving Position (see datasheet page 72, 8.2.25)
    -- db: Set partial driving to end at the last line
    -- (the datasheet says not to do this if you're not using it but whatever)
      0x34, 0x00, 0xdb,
      0x35, 0x00, 0x00,

    -- Horizontal and Vertical RAM Address Position
    -- (see datasheet page 74, 8.2.26)
    -- Set the window to the entire extents of the screen
      0x36, 0x00, 0xaf,
      0x37, 0x00, 0x00,
      0x38, 0x00, 0xdb,
      0x39, 0x00, 0x00,

    -- Gamma Control (see datasheet page 75, 8.2.27)
    -- (see also Gamma Correction, datasheet page 83, section 12)
    -- Set gamma fine adjustment for positive polarity output to 0x8a8800
      0x50, 0x00, 0x00,
      0x51, 0x08, 0x08,
      0x52, 0x08, 0x0a,
    -- Set gradient adjustment for positive polarity output to 0x0a
      0x53, 0x00, 0x0a,
    -- Set gamma fine adjustment for negative polarity output to 0x0088a8
      0x54, 0x0a, 0x08,
      0x55, 0x08, 0x08,
      0x56, 0x00, 0x00,
    -- Set gradient adjustment for negative polarity output to 0xa0
      0x57, 0x0a, 0x00,
    -- Set amplitude adjustment for positive polarity output to 0x710
      0x58, 0x07, 0x10,
    -- Set amplitude adjustment for negative polarity output to 0x710
      0x59, 0x07, 0x10,

    -- Display Control 1 again:
    -- 12: Ready the display to operate, but don't turn on
      0x07, 0x00, 0x12
    }
    -- Wait for the display operation to ready
    tmr.delay(50)
    -- 10: Enable tearing mitigation (frame sync)
    -- 17: Turn display on (and invert grayscale?)
    self:setreg{0x07, 0x10, 0x17}
  end

  function screen:fill(...)
    -- Select index register
    gpio.write(self.pins.rs, gpio.LOW)
    -- Select chip
    gpio.write(self.pins.cs, gpio.LOW)
    -- Send register index selection
    spi.send(1, 0, 0x22)

    -- Select data register
    gpio.write(self.pins.rs, gpio.HIGH)
    -- Send data bytes
    spi.send(1, ...)
    -- Deselect chip
    gpio.write(self.pins.cs, gpio.HIGH)
  end

  function screen:jump(x, y)
    self:setreg{
      0x20, 0x00, x,
      0x21, 0x00, y
    }
  end

  function screen:window(x0, x1, y0, y1, vertical)

    if vertical == nil then
      vertical = (x1 < x0) ~= (y1 < y0)
    end

    local modelo = vertical and 8 or 0

    if x1 < x0 then
      x0, x1 = x1, x0
    else
      modelo = modelo + 0x10
    end
    if y1 < y0 then
      y0, y1 = y1, y0
    else
      modelo = modelo + 0x20
    end

    self:setreg{
    -- Set window extents
      0x36, 0x00, x1,
      0x37, 0x00, x0,
      0x38, 0x00, y1,
      0x39, 0x00, y0,

    -- Set entry mode
      0x03, 0x10, modelo
    }
  end

  return screen
end
